Let's `#[expect]` some lints: Stabilize `lint_reasons` (RFC 2383) (#120924)
rust-lang/rust: src/doc/rustc/src/lints/levels.md

These lints exist mostly to be manually turned on via configuration, as we'll
talk about later in this section.

## expect

Sometimes, it can be helpful to suppress lints, but at the same time ensure that
the code in question still emits them. The 'expect' level does exactly this. If
the lint in question is not emitted, the `unfulfilled_lint_expectation` lint
triggers on the `expect` attribute, notifying you that the expectation is no
longer fulfilled.

```rust
fn main() {
    #[expect(unused_variables)]
    let unused = "Everyone ignores me";

>     #[expect(unused_variables)]
>     let used = "I'm useful";

Reviewer: nit: maybe add a comment in the code as to make it more obvious which `#[expect]` is unfulfilled
```suggestion
    #[expect(unused_variables)] // unused_variables lint is not emitted
    let used = "I'm useful";    // the expectation is therefore unfulfilled
```

    println!("The `used` value is equal to: {:?}", used);
}
```

This will produce the following warning:

```txt
warning: this lint expectation is unfulfilled
 --> src/main.rs:7:14
  |
7 |     #[expect(unused_variables)]
  |              ^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unfulfilled_lint_expectations)]` on by default
```

> This level can only be defined via the `#[expect]` attribute and not via the
> console. Lints with the special 'force-warn' lint will still be emitted as usual.

Reviewer: ```suggestion
This level can only be defined via the `#[expect]` attribute, there is no equivalent
flag. Lints with the special 'force-warn' level will still be emitted as usual.
```
